src/devices/cpu/pic16x8x/pic16x8x.cpp

```diff
@@ -0,0 +1,1443 @@
+	, m_data_config("data", ENDIANNESS_LITTLE, 8, 8, 0,
+			((pic_model == 0x83) ? address_map_constructor(FUNC(pic16x8x_device::ram_6), this) :
+			address_map_constructor(FUNC(pic16x8x_device::ram_7), this)))
+	, m_CONFIG(0x3fff)
+	, m_program_width(program_width)
+	, m_internal_eeprom_size(eeprom_size)
+	, m_read_port(*this, 0)
+	, m_write_port(*this)
+
+{
+}
+
+pic16cr83_device::pic16cr83_device(const machine_config &mconfig, const char *tag, device_t *owner, u32 clock)
+	: pic16x8x_device(mconfig, PIC16CR83, tag, owner, clock, 9, 0x40, 0x83)
+{
+}
+
+pic16cr84_device::pic16cr84_device(const machine_config &mconfig, const char *tag, device_t *owner, u32 clock)
+	: pic16x8x_device(mconfig, PIC16CR84, tag, owner, clock, 10, 0x40, 0x84)
+{
+}
+
+pic16f83_device::pic16f83_device(const machine_config &mconfig, const char *tag, device_t *owner, u32 clock)
+	: pic16x8x_device(mconfig, PIC16F83, tag, owner, clock, 9, 0x40, 0x83)
+{
+}
+
+pic16f84_device::pic16f84_device(const machine_config &mconfig, const char *tag, device_t *owner, u32 clock)
+	: pic16x8x_device(mconfig, PIC16F84, tag, owner, clock, 10, 0x40, 0x84)
+{
+}
+
+pic16f84a_device::pic16f84a_device(const machine_config &mconfig, const char *tag, device_t *owner, u32 clock)
+	: pic16x8x_device(mconfig, PIC16F84A, tag, owner, clock, 10, 0x40, 0x84)
+{
+}
+
+device_memory_interface::space_config_vector pic16x8x_device::memory_space_config() const
+{
+	return space_config_vector
+	{
+		std::make_pair(AS_PROGRAM, &m_program_config),
+		std::make_pair(AS_DATA,    &m_data_config)
+	};
+}
+
+std::unique_ptr<util::disasm_interface> pic16x8x_device::create_disassembler()
+{
+	return std::make_unique<pic16x8x_disassembler>();
+}
+
+
+/* Notes about NVRAM / PIC eeprom data
+
+Dump size is 4280 Bytes, 16 bits wide.
+0000h-03ffh is program data
+0400h-3fffh is dummy data
+4000h-400fh is user and config data
+4010h-41ffh is dummy data
+4200h-4280h is eeprom data
+
+NVRAM saved data is 8 bits wide, in accordance with the data width
+of the EEPROM memory of the PIC.
+This means that we need to perform the conversion from the default
+NVRAM dumped data to the NVRAM saved data.
+*/
+
+void pic16x8x_device::nvram_default()
+{
+	u16 dump_size = 0x4280;
+	u16 eeprom_dump = 0x4200;
+
+	// populate from a memory region if present
+	if (m_region.found())
+	{
+		if( m_region->bytes() != dump_size ) // pic memory dump total size
+		{
+			fatalerror( "Region '%s' wrong size (expected size = 0x%X)\n", tag(), dump_size );
+		}
+
+		if( m_region->bytewidth() != 2 ) // pic memory dumps are 16 bits wide
+		{
+			fatalerror( "Region '%s' needs to be an 16-bit region\n", tag() );
+		}
+
+		// Get default NVRAM data from memory region
+		memcpy(m_buff, m_region->base() + eeprom_dump, m_internal_eeprom_size * 2);
+
+		// Data width conversion
+		for (u8 i=0; i < m_internal_eeprom_size; i++)
+			m_eeprom_data[i] = m_buff[i] & 0x00ff;

```

Comment:
This code doesn't seem endian-safe.

Reply:
Since the rom load of this region is (and "must be") "ROM_REGION16_LE( 0x4280, "pic16f84", 0 )", I understand that the proposed code is correct.
The origin and destination data have been duly verified.
The memcpy() function loads u16 type data into m_buff from m_region, which in turn
have been read from the source in little_endian mode, therefore the conversion loop and its corresponding mask are correct.
Even so, if you consider that the code requires some change, I am completely willing to make it.

Reply:
Done the changes requested but missing the comments in the process. I'm sorry if I made some mistake.